src/steamship/agents/context.py

```diff
@@ -0,0 +1,39 @@
+from typing import Dict, Any, List, Optional
+
+from pydantic import BaseModel
+
+from steamship import Steamship
+from steamship.experimental.transports.chat import ChatMessage
+
+
+class ChatHistory(BaseModel):
+    chat_id: str = ""  # chat history is scoped by a chat_id. chat_id is unique across user_id's
+    messages: List[str] = []  # Represents a database that can select & sort messages
+
+    def get_history(self, message: ChatMessage, context):
+        return self.messages
+
+    def add_user_message(self, message: ChatMessage):
+        self.messages.append(f"user: {message}")
+
+    def add_ai_message(self, message: ChatMessage):
+        self.messages.append(f"AI: {message}")
+
+    def clear(self):
+        self.messages = []
+
+
+class AgentContext(BaseModel):
+    client: Steamship
+    chat_history: Optional[ChatHistory] = ChatHistory()
+    tracing: Optional[Dict[str, Any]] = []
+    scratchpad: Optional[List[str]] = []
+    new_message: ChatMessage
+
+    def emit(self, messages: List[ChatMessage]):
```

Comment:
Unsure if this is specifically what you were suggesting here, but FWIW I like the idea that there's an `emit` function (on the context?) that allows -- in theory -- anything to emit a message any time.

I also think that some standard contract of "Tools run, the agent observes their results, the agent responds" is useful, but just permitting the possibility that intermediate messages may be appropriate feels like a really great door to keep open.

Reply:
I agree. As a dev I felt comfortable with the notion that a Tool could just send a message to close the loop without having to return to the Agent. 

Flow that is now possible: 

Agent decides to execute long, asyc task -> Task gets executed -> Task ends by sending a message to user. 

VS 

Agent decides to execute long, asyc task -> Task gets executed -> Task ends by returning to Agent -> Agent has to know what to send to the user based on its context.